### PR TITLE
Backend Code Merge

### PR DESCRIPTION
**Pull request by ishranga11**
- Update backend code
- Adding exceptions handling

### CODE REVIEW

File: appengine/pom.xml [hunk 1]
```diff
@@ -91,6 +91,12 @@
             <artifactId>mysql-connector-java</artifactId>
             <version>8.0.19</version>
         </dependency>
+        <dependency>
```
Comment (dhruvpancholi) [inline appengine/pom.xml:94]: Dropwizard already comes with ObjectMapper, let's use that?

File: appengine/src/main/java/com/google/googleinterns/gscribe/GScribeConfiguration.java [hunk 2]
```diff
@@ -16,8 +16,44 @@
...
+     */
+    @Valid
+    @NotNull
+    private MySQLConfig mySQLConfig;
+
+    @Valid
+    @NotNull
+    private Environment environment;
+
+    @JsonProperty("database")
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/GScribeConfiguration.java:40]: Why is this required?

File: appengine/src/main/java/com/google/googleinterns/gscribe/config/MySQLConfig.java [hunk 3]
```diff
@@ -17,14 +17,11 @@
...
 
 public class MySQLConfig {
-    private final String url;
-    private final String userName;
-    private final String password;
+    private String url;
+    private String userName;
+    private String password;
 
-    public MySQLConfig(String url, String userName, String password) {
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/config/MySQLConfig.java:24]: How are the variables set? Is it getting deserialized?

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+public interface AnswerDao {
+
+    /**
+     * Queries list of all answers for a particular exam instance identified by an exam instance id examInstanceID
+     *
+     * @param examInstanceID ( to identify particular exam instance )
+     * @return List of answers
+     */
+    @Mapper(AnswerDao.AnswerMapper.class)
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java:40]: AnswerDao.AnswerMapper.class -> AnswerMapper.class

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    @SqlQuery("SELECT * from answers where exam_instance_id = :exam_instance_id")
+    List<Answer> getAnswersByExamInstanceID(@Bind("exam_instance_id") int examInstanceID);
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java:42]: Instead of List<Answer>, simply return Answer.

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     * Called to insert answers into the database for an exam instance ID
+     * Inserts All responses for a corresponding exam instance
+     *
+     * @param examInstanceID ( to identify particular exam instance )
+     * @param questionNumber ( question number of corresponding answer )
+     * @param answers        ( answer JSON )
+     */
+    @SqlBatch("INSERT INTO answers ( exam_instance_id, answer, question_num ) VALUES ( :exam_instance_id, :answer, :question_num )")
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java:52]: We decided to add all the answers in one row itself.

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    void insertAnswers(@Bind("exam_instance_id") int examInstanceID, @Bind("question_num") List<Integer> questionNumber, @Bind("answer") List<String> answers);
+
+    /**
+     * A Mapper class to map answer JSON object from MySQL database to Answer class
+     */
+    class AnswerMapper implements ResultSetMapper<Answer> {
+        @Override
+        public Answer map(int i, ResultSet resultSet, StatementContext statementContext) throws SQLException {
+            return new Gson().fromJson(resultSet.getString("answer"), Answer.class);
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java:61]: Use ObjectMapper?

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/ExamInstanceDao.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+     * Called to check if a examinee identified by roll number rollNumber already attempted the exam represented by examID
+     * Queries exam instance using exam id and student roll number
+     *
+     * @param examID     ( to identify particular exam )
+     * @param rollNumber ( roll number of examinee )
+     * @return exam instance object
+     */
+    @Mapper(ExamInstanceDao.ExamInstanceMapper.class)
+    @SqlQuery("SELECT * from exam_instance where exam_id = :exam_id and student_roll_num = :student_roll_num")
+    ExamInstance getExamInstanceByUserDetails(@Bind("exam_id") int examID, @Bind("student_roll_num") int rollNumber);
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/dao/ExamInstanceDao.java:63]: Add email ID as well.

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/ExamMetadataDao.java [hunk 9]
```diff
@@ -16,5 +16,83 @@
...
+    ExamMetadata getExamMetadataByExamId(@Bind("exam_id") int examID);
+
+    /**
+     * Called to populate exam metadata table on paper setter dashboard
+     * Queries list of exam metadata of exams created by user identified by userID
+     *
+     * @param userID ( unique user id of user, here paper setter )
+     * @return list of exam metadata
+     */
+    @Mapper(ExamMetadataDao.ExamMetadataMapper.class)
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/dao/ExamMetadataDao.java:65]: Same

**Review by dhruvpancholi — CHANGES_REQUESTED** [review]

File: appengine/pom.xml (hunk 1 above)
Comment (ishranga11) [reply appengine/pom.xml:94]: Updated code to use ObjectMapper

File: appengine/src/main/java/com/google/googleinterns/gscribe/GScribeConfiguration.java (hunk 2 above)
Comment (ishranga11) [reply appengine/src/main/java/com/google/googleinterns/gscribe/GScribeConfiguration.java:40]: This is used to read database credentials from the YAML file itself.

File: appengine/src/main/java/com/google/googleinterns/gscribe/config/MySQLConfig.java (hunk 3 above)
Comment (ishranga11) [reply appengine/src/main/java/com/google/googleinterns/gscribe/config/MySQLConfig.java:24]: Forgot to add a new MySQLConfig file in this PR. This will be removed and a normal constructor for deserialization is used there. Will update the PR to add a new MySQLConfig file.

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java (hunk 4 above)
Comment (ishranga11) [reply appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java:40]: Updated.

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java (hunk 7 above)
Comment (ishranga11) [reply appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java:61]: Updated.

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/ExamMetadataDao.java (hunk 9 above)
Comment (ishranga11) [reply appengine/src/main/java/com/google/googleinterns/gscribe/dao/ExamMetadataDao.java:65]: Updated.

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java (hunk 6 above)
Comment (ishranga11) [reply appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java:52]: Followed the procedure used in inserting questions into the database as ( exam_id, question_number, question_JSON ). Should this be changed to a single JSON entry for the whole exam instance?
